x/ecocredit/server/marketplace/buy.go

```diff
@@ -2,10 +2,255 @@ package marketplace
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
-func (k Keeper) Buy(ctx context.Context, req *marketplace.MsgBuy) (*marketplace.MsgBuyResponse, error) {
-	panic("not implemented")
+// Buy allows users to purchase credits by either directly specifying a sell order, or
+// defining a set of filters with attributes to match against.
+//
+// Currently, only the former is supported. Calls to this function with anything other than
+// MsgBuy_Order_Selection_SellOrderId will fail.
+func (k Keeper) Buy(ctx context.Context, req *v1.MsgBuy) (*v1.MsgBuyResponse, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	buyerAcc, err := sdk.AccAddressFromBech32(req.Buyer)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, order := range req.Orders {
+		// assert they have the balance they're  bidding with
+		bal := k.bankKeeper.GetBalance(sdkCtx, buyerAcc, order.BidPrice.Denom)
+		if bal.IsLT(*order.BidPrice) {
+			return nil, sdkerrors.ErrInsufficientFunds.Wrapf("cannot bid %v with a balance of %v", bal, *order.BidPrice)
+		}
+
+		switch selection := order.Selection.Sum.(type) {
+		case *v1.MsgBuy_Order_Selection_SellOrderId:
+			sellOrder, err := k.stateStore.SellOrderTable().Get(ctx, selection.SellOrderId)
+			if err != nil {
+				return nil, fmt.Errorf("sell order %d: %w", selection.SellOrderId, err)
+			}
+			if sellOrder.DisableAutoRetire != order.DisableAutoRetire {
+				return nil, sdkerrors.ErrInvalidRequest.Wrapf("auto-retire mismatch: sell order set to %t, buy "+
+					"order set to %t", sellOrder.DisableAutoRetire, order.DisableAutoRetire)
+			}
+			batch, err := k.coreStore.BatchInfoTable().Get(ctx, sellOrder.BatchId)
+			if err != nil {
+
+				return nil, sdkerrors.ErrIO.Wrapf("error getting batch id %d: %s", sellOrder.BatchId, err.Error())
+			}
+			ct, err := utils.GetCreditTypeFromBatchDenom(ctx, k.coreStore, k.paramsKeeper, batch.BatchDenom)
+			if err != nil {
+				return nil, err
+			}
+			creditOrderQty, err := math.NewPositiveFixedDecFromString(order.Quantity, ct.Precision)
+			if err != nil {
+				return nil, err
+			}
+
+			// check that bid price and ask price denoms match
+			market, err := k.stateStore.MarketTable().Get(ctx, sellOrder.MarketId)
+			if err != nil {
+				return nil, fmt.Errorf("market id %d: %w", sellOrder.MarketId, err)
+			}
+			if order.BidPrice.Denom != market.BankDenom {
+				return nil, sdkerrors.ErrInvalidRequest.Wrapf("bid price denom does not match ask price denom: "+
+					" %s, expected %s", order.BidPrice.Denom, market.BankDenom)
+			}
+
+			// check that bid price is at least equal to ask price
+			askAmount, ok := sdk.NewIntFromString(sellOrder.AskPrice)
+			if !ok {
+				return nil, fmt.Errorf("could not convert sell order's ask price to %T: %s", sdk.Int{}, sellOrder.AskPrice)
+			}
+			if order.BidPrice.Amount.LT(askAmount) {
+				return nil, sdkerrors.ErrInsufficientFunds.Wrapf("bid price too low: got %s, ask: %s",
```

Comment:
insufficient funds isn't the right error

Reply:
added custom error